src/datasets/utils/metadata.py

```diff
@@ -106,6 +111,68 @@ def to_yaml_string(self) -> str:
         ).decode("utf-8")
 
 
+class MetadataConfigs(Dict[str, dict]):
+    """Should be in format {config_name: {**config_params}}."""
+
+    __configs_field_name: ClassVar[str] = METADATA_CONFIGS_FIELD
+
+    @classmethod
+    def from_metadata(cls, dataset_metadata: DatasetMetadata) -> "MetadataConfigs":
+        if cls.__configs_field_name in dataset_metadata:
+            metadata_configs = dataset_metadata.get(cls.__configs_field_name)
+            if isinstance(metadata_configs, dict):
+                config_name = metadata_configs.pop("config_name", "default")
+                return cls({config_name: metadata_configs})
+            if isinstance(metadata_configs, list):
+                return cls(
+                    {
+                        config["config_name"]: {
+                            param: value for param, value in config.items() if param != "config_name"
+                        }
+                        for config in metadata_configs
+                    }
+                )
+        return cls()
+
+    def to_metadata(self, dataset_metadata: DatasetMetadata) -> None:
+        if self:
+            current_metadata_configs = self.from_metadata(dataset_metadata)
+            total_metadata_configs = dict(sorted({**current_metadata_configs, **self}.items(), key=lambda x: x[0]))
+            if len(total_metadata_configs) > 1:
+                for config_name, config_metadata in total_metadata_configs.items():
+                    config_metadata.pop("config_name", None)
+                dataset_metadata[self.__configs_field_name] = [
+                    {"config_name": config_name, **config_metadata}
+                    for config_name, config_metadata in total_metadata_configs.items()
+                ]
+            elif len(total_metadata_configs) == 1:
+                metadata_config_name, metadata_config = next(iter(total_metadata_configs.items()))
+                metadata_config = (
+                    {"config_name": metadata_config_name, **metadata_config}
+                    if metadata_config_name != "default"
+                    else {**metadata_config}
+                )
+                dataset_metadata[self.__configs_field_name] = metadata_config
+
+    def to_builder_configs_list(self, builder_config_cls):
+        metadata_configs = copy.deepcopy(self)
+        return [
+            builder_config_cls(
+                name=name,
+                **{param: value for param, value in meta_config.items() if hasattr(builder_config_cls, param)},
+            )
+            for name, meta_config in metadata_configs.items()
```

Comment:
One shouldn't call `to_builder_configs_list` without resolving the data_files. Could you raise an error if it's the case and suggest calling e.g. `get_data_files_for_metadata_configs_locally ` ?

`get_data_files_for_metadata_configs_locally ` and `get_data_files_for_metadata_configs_in_dataset_repository` can be methods of `MetadataConfigs`, maybe
```python
metadata_configs.resolve_data_files_locally(base_path, with_metadata_files)
```
and
```python
metadata_configs.resolve_data_files_in_dataset_repository(hfh_info, base_path, with_metadata_files)
```

Reply:
[cf293e4](https://github.com/huggingface/datasets/pull/5331/commits/cf293e4ad615547c776a85ddbc3e729873ad4098)